File: 103. Binary Tree Zigzag Level Order Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        if (root == NULL) return {};
        queue <TreeNode*> qnode;
        queue <int> qlevel;
        stack <TreeNode*> snode;
        stack <int> slevel;
        qnode.push(root);
        qlevel.push(0);
        vector<vector<int>> res {{}};
        while (!qnode.empty())
        {
            root = qnode.front(); qnode.pop();
            int level = qlevel.front(); qlevel.pop();
            res[res.size()-1].push_back(root->val);
            if (level % 2 == 0)
            {
                if (root->left)
                {
                    snode.push(root->left);
                    slevel.push(level+1);
                }
                if (root->right)
                {
                    snode.push(root->right);
                    slevel.push(level+1);
                }
            }
            else
            {
                if (root->right)
                {
                    snode.push(root->right);
                    slevel.push(level+1);
                }
                if (root->left)
                {
                    snode.push(root->left);
                    slevel.push(level+1);
                }
            }
            if (qnode.empty() && !snode.empty())
            {
                while (!snode.empty())
                {
                    qnode.push(snode.top()); snode.pop();
                    qlevel.push(slevel.top()); slevel.pop();
                }
            }
            if (!qnode.empty() && level < qlevel.front())
                res.push_back({});
        }
        return res;
    }
};

File: 103. Binary Tree Zigzag Level Order Traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def zigzagLevelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        if not root:
            return []
        queue = [(root, 0)]
        res = [[]]
        stack = []
        while queue:
            node, level = queue.pop(0)
            res[len(res)-1] += [node.val]
            if level % 2 ==0:
                if node.left:
                    stack += [(node.left, level+1)]
                if node.right:
                    stack += [(node.right, level+1)]
            else:
                if node.right:
                    stack += [(node.right, level+1)]
                if node.left:
                    stack += [(node.left, level+1)]
            if not queue and stack:
                while stack:
                    queue += [stack.pop()]
            if  queue and level < queue[0][1]:
                res += [[]]
        return res
